plugins/completion/nvim-cmp/cmp-helpers.nix

```diff
@@ -1,54 +1,119 @@
+  removeNull = list:
+    if list == [] then
+      []
+    else if isNull (head list) then
+      [] ++ removeNull (drop 1 list)
+    else
+      [ (head list) ] ++ removeNull (drop 1 list);
+in with helpers;
+rec {
+  # mkCmpSourcePlugin = { name, extraPlugins ? [], useDefaultPackage ? true, ... }: mkPlugin attrs {
+  #   inherit name;
+  #   extraPlugins = extraPlugins ++ (lists.optional useDefaultPackage pkgs.vimExtraPlugins.${name});
+  #   description = "Enable ${name}";
+  # };
+
+  # value is string (package name) or a set of form '{ package = "..."; extraConfig = "..."; }'
+  # the former is needed if source needs extra config like the crates-nvim source
+  sourceNameAndPlugin = {
     "buffer" = "cmp-buffer";
     "calc" = "cmp-calc";
-    "dictionary" = "cmp-dictionary";
-    "digraphs" = "cmp-digraphs";
-    "omni" = "cmp-omni";
-    "spell" = "cmp-spell";
-    "nvim_lsp" = "cmp-nvim-lsp";
-    "nvim_lsp_document_symbol" = "cmp-nvim-lsp-document-symbol";
-    "nvim_lsp_signature_help" = "cmp-nvim-lsp-signature-help";
-    "vim_lsp" = "cmp-vim-lsp";
-    "path" = "cmp-path";
-    "git" = "cmp-git";
-    "conventionalcommits" = "cmp-conventionalcommits";
     "cmdline" = "cmp-cmdline";
-    "cmp-cmdline-history" = "cmp-cmdline-history";
-    "fuzzy_buffer" = "cmp-fuzzy-buffer";
-    "fuzzy_path" = "cmp-fuzzy-path";
-    "rg" = "cmp-rg";
-    "fish" = "cmp-fish";
-    "tmux" = "cmp-tmux";
-    "zsh" = "cmp-zsh";
-    "crates" = "crates-nvim";
-    "npm" = "cmp-npm";
     "cmp-clippy" = "cmp-clippy";
+    "cmp-cmdline-history" = "cmp-cmdline-history";
+    "cmp_pandoc" = "cmp-pandoc-nvim";
     "cmp_tabnine" = "cmp-tabnine";
+    "conventionalcommits" = "cmp-conventionalcommits";
     "copilot" = "cmp-copilot";
+    "crates" = { package = "crates-nvim"; extraConfig = "require('crates').setup()"; };
     "dap" = "cmp-dap";
+    "dictionary" = "cmp-dictionary";
+    "digraphs" = "cmp-digraphs";
     "emoji" = "cmp-emoji";
+    "fish" = "cmp-fish";
+    "fuzzy_buffer" = "cmp-fuzzy-buffer";
+    "fuzzy_path" = "cmp-fuzzy-path";
+    "git" = "cmp-git";
     "greek" = "cmp-greek";
     "latex_symbols" = "cmp-latex-symbols";
     "look" = "cmp-look";
+    "luasnip" = "cmp-luasnip";
+    "npm" = "cmp-npm";
+    "nvim_lsp" = "cmp-nvim-lsp";
+    "nvim_lsp_document_symbol" = "cmp-nvim-lsp-document-symbol";
+    "nvim_lsp_signature_help" = "cmp-nvim-lsp-signature-help";
     "nvim_lua" = "cmp-nvim-lua";
+    "omni" = "cmp-omni";
     "pandoc_references" = "cmp-pandoc-references";
-    "cmp_pandoc" = "cmp-pandoc-nvim";
+    "path" = "cmp-path";
+    "rg" = "cmp-rg";
+    "snippy" = "cmp-snippy";
+    "spell" = "cmp-spell";
+    "tmux" = "cmp-tmux";
     "treesitter" = "cmp-treesitter";
+    "ultisnips" = "cmp-nvim-ultisnips";
+    "vim_lsp" = "cmp-vim-lsp";
     "vimwiki-tags" = "cmp-vimwiki-tags";
+    "vsnip" = "cmp-vsnip";
+    "zsh" = "cmp-zsh";
   };
+
+  # create a list off all packages of all activated sources for extraPackgs
```

Comment:
Sorry to be pedantic :sweat_smile: but of*, I'd be changing things here anyway so doesn't hurt.

Reply:
I am not sure what you mean. What do you want me to change here?

Reply:
off -> of, just spelling :sweat_smile: